### ISSUE TITLE
Freeze when calling Disconnect() on Client

### ISSUE THREAD

**Issue by LeaT113** [jchristn/CavemanTcp#4]
I'm having issue disconnecting clients from a server. The server can disconnect them just fine, but when I call Disconnect() on the client, it never returns. Same with the Dispose(), which seems to be basically equivalent.

Edit: I have created a very basic server / client app with CavemanTcp and there the Disconnect on Client works. But I'm still unable to fix my issue in my actual code. From looking at the Dispose method in you code, I'm unable to think of a reason it would not return. Any help would be appreciated.

**Comment by jchristn** [comment]
Hi @LeaT113 which OS and framework are you using?  Do you experience the same behavior with the Test.Client and Test.Server projects?

I just tried on Windows 10 with .NET 5.0, .NET Core 3.1, and .NET Framework 4.6.1 and could not reproduce.  If you could please let me know OS/framework and share any code that could help me reproduce.  Thanks!

**Comment by LeaT113** [comment]
Hi @jchristn, thanks for the timely response. 

I'm using Windows 10 (20H2) and the project is on .NET Framework 4.7.2 and the WPF platform. 
In my very simple WPF test app, where I've only created a server and client and connect/disconnect via buttons, the client can call the Disconnect method just fine. This is one app that implements both server and client.

But in my app, which is [here on github](https://github.com/LeaT113/TTransfer/tree/exceptions), it seems to just freeze. The code I'm mentioning is on the 'exceptions' branch.

You can check my [client code right here](https://github.com/LeaT113/TTransfer/blob/exceptions/TTransfer/Network/TransferClient.cs). Specifically, it's the TerminateConnection method, which is responsible for killing the connection in the event of a problem.

I have tried calling it via Invoke on the UI thread as I thought maybe there was some thread ownership problem, but that made no difference.

**Comment by jchristn** [comment]
Thank you @LeaT113 - good to know.  Everything looks really clean there.  Would you by chance be in the midst of a read or write (perhaps some background task) when ```Disconnect()``` is called?  Reason I ask is ```Disconnect``` calls ```Dispose``` which waits until the read/write semaphores can be entered.  That can create a blocking situation.

**Comment by LeaT113** [comment]
I noticed the semaphores, but I'm unfamiliar with them. I do not believe there's any reading or sending taking place. All the reading on my client is done via the `TryRead` method that throws an exception if it doesn't receive data in time, which is at max 10s. I have waited longer and the app stays frozen after calling `Disconnect`.

I just did a test. I made the server purposely not send anything. What happens is, the client uses your `Connect`. In the `Events_ClientConnected`, it runs an async method `HandleConnection`. That runs async `EstablishConnection` and in that, I use the aforementioned `TryRead` method, which internally uses your `ReadWithTimeout`. I have checked it then throws an exception, which is caught inside the `HandleConnection` method and that method calls `TerminateConnection`. I know that to be true. 
Also in this test, no sending is even initiated and the only reading is done via the one `TryRead` method, that throws the exception.

For some reason, this process causes the error. If I just manually call your `Disconnect` method via a button, it completes just fine. I thought maybe it has something to do with the async methods, but I'm new to async and I'm not sure exactly what problems could arise, since I even tried invoking it on the UI thread.
I have also put a try catch block around the `Disconnect` and it did not throw an exception.

**Comment by jchristn** [comment]
Thanks @LeaT113.  Sounds like there are two issues.

1) There's an exception you're getting from ```ReadWithTimeout``` - could you capture that and paste it here?  There shouldn't be anything in ```ReadWithTimeout``` that propagates an exception up to your code.  

2) IsConnected is not properly reflecting that the client is connected.  I think this can be fixed pretty easily; right now in the ```Connect``` method, the event is fired before the variable is set to ```true```.  I'm doing a patch for that right now.

**Comment by jchristn** [comment]
v1.3.3.2: https://www.nuget.org/packages/CavemanTcp/1.3.3.2 - this should resolve the ```IsConnected``` property.
Commit: https://github.com/jchristn/CavemanTcp/commit/a44be55be831e95ad727989c5f08097318460137

Please let me know about the exception when you have a moment.  Cheers, Joel

**Comment by LeaT113** [comment]
Hi @jchristn.

I may have not explained it very well. I'm not getting any exceptions from `ReadWithTimeout`. I'm using custom exceptions. You can look into the `TryRead` method inside my `TransferClient` class for details, but it's simple. If the `ReadResultStatus` of the read operation is not `Success`, I raise my exception. So there's no exception coming from your code.

But I have been able to reproduce the issue in my test app now. The problem is calling `ReadWithTimeout`. I call it, I give it a timeout of 1 second. The server doesn't send anything, so after that 1 second, it returns with the `ReadResultStatus` of `Timeout`. That part works correctly. But calling `Disconnect` then will produce the freezing issue. If I remove the `ReadWithTimeout` call, it works as intended. So what I'm guessing is, in the even of a Timeout, the `ReadWithTimeout` doesn't correctly end the semaphore. 

I have created a repo for the test app, you can [find it here](https://github.com/LeaT113/CavemanServer). There are instructions on how to use it, it's a bit rough, but works. You might need to expand the window a tiny bit to get rid of the Control clipping, not sure why that happens.

Thanks, Tomáš

**Comment by jchristn** [comment]
Got it, thank you for making it clear. I will dig into this ASAP!  Cheers, Joel

**Comment by jchristn** [comment]
By the way, did 1.3.3.2 fix the ```IsConnected``` issue?

**Comment by jchristn** [comment]
Was able to reproduce, working on this now.

**Comment by LeaT113** [comment]
Yes, as of 1.3.3.2, the `IsConnected` bool is `true` right at the start of the `ClientConnected` event. Works perfectly.

I'm glad you're able to reproduce it and I'm really thankful about how enthusiastically you're working on figuring this out. Thanks.

**Comment by jchristn** [comment]
Happy to.  The read semaphore isn't being released properly; that's the core of the problem.  Will update as soon as I have a fix.

**Comment by jchristn** [comment]
Could you give this a go and see if this fixes the issue?  Might take a few mins before the package is available so you can update.

NuGet: https://www.nuget.org/packages/CavemanTcp/1.3.3.3
Commit: https://github.com/jchristn/CavemanTcp/commit/eb83531e9cef4ec0ff1752d390347d7664483d59

**Comment by LeaT113** [comment]
That seems to have solved the issue in the test app! I'll check the real app now.

**Comment by LeaT113** [comment]
Yes, that works perfectly now! Thank you so much.
And thanks for such quick responses. I might try to contribute to the documentation after finishing my project, as I really appreciate CavemanTcp. 

Just one thing, is there a reason why the summary xml is not shipped with the package via NuGet? I had to manually download it from the repo and paste it into my project to get method summaries.

**Comment by jchristn** [comment]
Ooooh - it should definitely be in there. I have a couple more items to do and will make sure the XML is in there in the next patch. 

Thanks for letting me know about this issue!  I'll close it out when I do the next package in the next day or so.

**Comment by LeaT113** [comment]
I'm not happy to raise another problem, but I may have run into another issue.

I'm using `client.Send` to send parts of a file streamed from disk. The overall loop is very fast (and gets run many times, my files are segmented into packets of max 1024B), uses very little of my CPU (R3600, max around 12%). A 150MB file gets sent in average 3.5s. 
But I replaced the `client.Send` with `client.SendWithTimeout` to catch any issues during the file transfer and terminate it. But that change dramatically affects transfer speeds, now using about 25% of my CPU and taking on average 13.6s.

**Comment by jchristn** [comment]
Yes, ```SendWithTimeout``` is much much harder on the CPU.  Each time you call it, it starts two tasks.  The first containing the attempt to send the data, and the second, a timeout task.  After either your send completes, or, the timeout is exceeded, both tasks have to be canceled.  

Out of curiosity, have you considered using a library that has built-in framing?  I am also publisher of WatsonTcp (https://github.com/jchristn/watsontcp) which makes things like this really easy. 

CavemanTcp is primarily for building your own state machines or your own framing layers.

**Comment by jchristn** [comment]
Another thing you can do is use Send and the send the final block in SendWithTimeout

**Comment by LeaT113** [comment]
Hmm, that makes sense. Question is if I need to `Send` with a timeout, I'm not sure exactly in which scenarios that would actually time out. I noticed some CancellationTokens in the code, but couldn't figure out a way to use them, so I assume there's no way to manually cancel a `Send` or `Read` operation from CavemanTcp.

It's also important for me to stop the file transfer as soon as possible when something fails, since I don't want the client to be sending potentially gigabytes of a file that is not gonna be received anyway.

I have checked out your entire impressive repertoire before settling on CavemanTcp. I have already written a more basic form of TTransfer using just the native `TcpClient` and `TcpServer` classes, so I wanted something as simple and reasonably low level as I could find. I thought it would also aid in optimizing for the fastest transfers, since I knew I'd be streaming files. 

From looking at WatsonTcp, it doesn't seem extremely different, mainly having events for receiving, at least from a quick look. And the framing. I'm gonna be honest, I don't know much about framing. But I haven't run into any issues using the previous version of TTransfer, which ran on just the native client and server. My app is meant to be used in a local network, but it did pass testing over an internet VPN tunnel as well. 
I also do a form of state machine while establishing a connection, so I need a lot of manual control. I think at least. I'm no TCP expert.

**Comment by jchristn** [comment]
Framing is really important if you are sending variable-length messages.  While TCP will ensure an in-order delivery of the data sent (segments are organized according to their sequence number in the receiver's buffer prior to the application retrieving them), the receiver has to read some number of bytes.  If the receiver doesn't know the number to read, they may read too many, too few, etc.  i.e. the frame is really nothing more than a wrapper that the receiver examines first to determine the number of actual bytes to read.  In WatsonTcp you can also add metadata to it, i.e. send your own objects across in the frame itself, which can be useful to the reader.  Stephen Cleary wrote a really in-depth blog post about it some time ago which I still use as a reference: https://blog.stephencleary.com/2009/04/message-framing.html

One of the side effects (positive) of having framing is that you can have an event driven implementation where the library will fire an event when an entire message is received (which is what WatsonTcp does).  As an aside, the guts/internals of Watson are very similar to CavemanTcp, but what is exposed to consuming applications is curated to provide an easier experience for the developer.

In your case, it sounds like you have fixed-width messages and are doing your own state machine during connection setup, so I don't think that you would need something like what WatsonTcp is providing.

Timeouts could realistically occur anywhere due to packet loss, congestion, intermediary devices failing and new routes being used, etc.  For your case, if you have fixed-width messages, I assume part of each message is for control purposes and the other part for data.  It may be worthwhile to have the receiver send checkpoints back to the sender using the control part of the message to relay back their perspective of the current transfer.  That could also be a clue to indicate whether or not the transfer should continue or be stopped.

The cancellation tokens are primarily there to allow the operation to be aborted by the caller.  You're correct that ```Send``` and ```Read``` can't be canceled, but the async methods can.  The easiest way to do this would be to instantiate a ```CancellationToken``` outside of the ```SendAsync``` or ```ReadAsync``` call, and pass it into that method, i.e.
```csharp
CancellationTokenSource cts = new CancellationTokenSource();
CancellationToken ct = cts.Token;

WriteResult wr = client.SendAsync("hello", ct);
```
and then somewhere else...
```
if (some condition is met where you want to cancel)
{
  cts.Cancel(); // kill the send
}
```

**Comment by jchristn** [comment]
Hi @LeaT113 wanted to let you know I just published v1.3.4 which has some good fixes around cancellation tokens as well as the XML documentation.  Going to close this one out; thank you so much for your help identifying this, and I hope it's ok I added you to the README.

Cheers, Joel

NuGet: https://www.nuget.org/packages/CavemanTcp/1.3.4
Commit: https://github.com/jchristn/CavemanTcp/commit/488fb11c67cc8ef9da1ce1c28a120d5bdd554dec
